src/Lifecycle.js

```diff
@@ -348,15 +348,26 @@ export function setLoggedIn(credentials) {
  * new one in its place. This additionally starts all other react-sdk services
  * which use the new Matrix client.
  *
+ * If the credentials belong to a different user from the session already stored,
+ * the old session will be cleared automatically.
+ *
  * @param {MatrixClientCreds} credentials The credentials to use
  *
  * @returns {Promise} promise which resolves to the new MatrixClient once it has been started
  */
 export function hydrateSession(credentials) {
-    stopMatrixClient();
+    const oldUserId = MatrixClientPeg.get().getUserId();
+
+    stopMatrixClient(); // unsets MatrixClientPeg.get()
     localStorage.removeItem("mx_soft_logout");
     _isLoggingOut = false;
-    return _doSetLoggedIn(credentials, false);
+
+    const overwrite = credentials.userId !== oldUserId;
```

Comment:
the reasoning here is that through SSO the user could pick another user? Because through password login, they should not be able to...

Reply:
SSO is the easiest way to trigger this, although technically speaking password login *can* result in a different user ID however that would be a very strange implementation choice.